File: prime_finder.c
// prime_finder.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <stdint.h>

typedef uint64_t u64;
typedef uint32_t u32;

void sieve_basic(u32 limit, u32 *primes, u32 *count) {
    char *is_prime = calloc(limit + 1, sizeof(char));
    memset(is_prime, 1, limit + 1);
    is_prime[0] = is_prime[1] = 0;
    
    for (u32 i = 2; i * i <= limit; i++) {
        if (is_prime[i]) {
            for (u32 j = i * i; j <= limit; j += i) {
                is_prime[j] = 0;
            }
        }
    }
    
    *count = 0;
    for (u32 i = 2; i <= limit; i++) {
        if (is_prime[i]) {
            primes[(*count)++] = i;
        }
    }
    free(is_prime);
}

u64 segmented_sieve(u64 limit) {
    u32 sqrt_limit = (u32)sqrt(limit) + 1;
    u32 *base_primes = malloc(sqrt_limit * sizeof(u32));
    u32 base_count;
    
    sieve_basic(sqrt_limit, base_primes, &base_count);
    
    u64 count = base_count;
    u64 segment_size = 1000000; // 1M segment
    char *segment = malloc(segment_size);
    
    for (u64 low = sqrt_limit + 1; low <= limit; low += segment_size) {
        u64 high = (low + segment_size - 1 < limit) ? low + segment_size - 1 : limit;
        memset(segment, 1, high - low + 1);
        
        for (u32 i = 0; i < base_count; i++) {
            u32 p = base_primes[i];
            u64 start = ((low + p - 1) / p) * p;
            if (start < low) start += p;
            if (start == p) start += p; // Skip the prime itself
            
            for (u64 j = start; j <= high; j += p) {
                segment[j - low] = 0;
            }
        }
        
        for (u64 i = 0; i <= high - low; i++) {
            if (segment[i]) count++;
        }
    }
    
    free(base_primes);
    free(segment);
    return count;
}

int main() {
    clock_t start = clock();
    u64 result = segmented_sieve(10000000000ULL); // 10^10
    clock_t end = clock();
    
    printf("Primes up to 10^10: %llu\n", result);
    printf("Time: %.2f seconds\n", ((double)(end - start)) / CLOCKS_PER_SEC);
    return 0;
}

File: prime_finder.go
// prime_finder.go
package main

import (
    "fmt"
    "math"
    "time"
)

func sieveBasic(limit uint32) []uint32 {
    isPrime := make([]bool, limit+1)
    for i := range isPrime {
        isPrime[i] = true
    }
    isPrime[0], isPrime[1] = false, false
    
    for i := uint32(2); i*i <= limit; i++ {
        if isPrime[i] {
            for j := i * i; j <= limit; j += i {
                isPrime[j] = false
            }
        }
    }
    
    var primes []uint32
    for i := uint32(2); i <= limit; i++ {
        if isPrime[i] {
            primes = append(primes, i)
        }
    }
    return primes
}

func segmentedSieve(limit uint64) uint64 {
    sqrtLimit := uint32(math.Sqrt(float64(limit))) + 1
    basePrimes := sieveBasic(sqrtLimit)
    
    count := uint64(len(basePrimes))
    segmentSize := uint64(1000000)
    segment := make([]bool, segmentSize)
    
    for low := uint64(sqrtLimit) + 1; low <= limit; low += segmentSize {
        high := low + segmentSize - 1
        if high > limit {
            high = limit
        }
        
        for i := uint64(0); i <= high-low; i++ {
            segment[i] = true
        }
        
        for _, p := range basePrimes {
            p64 := uint64(p)
            start := ((low + p64 - 1) / p64) * p64
            if start < low {
                start += p64
            }
            if start == p64 {
                start += p64
            }
            
            for j := start; j <= high; j += p64 {
                segment[j-low] = false
            }
        }
        
        for i := uint64(0); i <= high-low; i++ {
            if segment[i] {
                count++
            }
        }
    }
    
    return count
}

func main() {
    start := time.Now()
    result := segmentedSieve(10000000000)
    duration := time.Since(start)
    
    fmt.Printf("Primes up to 10^10: %d\n", result)
    fmt.Printf("Time: %.2f seconds\n", duration.Seconds())
}